support for Ipv6

BPplays opened the pull request:
<!-- Put X in the box below to confirm -->
this pr adds support for ipv6
it changes the autorun url to use `localhost` instead of `127.0.0.1` and adds a separate step to start listening on v6 and has error handling and should still work even if it can't start either v6 or v4, but it exits if both v6 and v4 fail to start.

it also changes `dns.setDefaultResultOrder` to `ipv6first` im not sure why `dns.setDefaultResultOrder` was set to `ipv4first`, [release 1.10.3](https://github.com/SillyTavern/SillyTavern/releases/tag/1.10.3) claims it to either fix a bug but doesn't say what the bug is, or that it is a bug. i didn't have any issues with it in testing
​

btw i don't have much experience with node

​
​
​
​
​
## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

BPplays committed:
started adding v6 support

BPplays committed:
added error checking and change messages to the user

BPplays committed:
fixed lsp caused issue

BPplays committed:
fixed formatting error

BPplays committed:
added error handling to https

BPplays committed:
fixed formatting errors

Cohee1207 commented:
> it also changes dns.setDefaultResultOrder to ipv6first im not sure why dns.setDefaultResultOrder was set to ipv4first, [release 1.10.3](https://github.com/SillyTavern/SillyTavern/releases/tag/1.10.3) claims it to either fix a bug but doesn't say what the bug is, or that it is a bug. i didn't have any issues with it in testing

Because many ISPs still have issues with IPv6, and it used to cause issues when the resolved IP was not available to the client because it defaulted to IPv6. Even if you don't experience problems with it, someone else will. So please put it back as it was.

Cohee1207 commented on server.js:
```diff
@@ -708,22 +714,68 @@ userModule.initUserStorage(dataRoot)
-                .listen(
-                    Number(tavernUrl.port) || 443,
+            // Handle IPv6 server
+            try {
```
This is not readable. Extract server creation into a function.

Cohee1207 commented:
This should be enabled as a config value with opt-in. I don't want to force this on everyone.

Wolfsblvt commented:
Could just be a yaml switch to enable (and maybe another to default to) ipv6, like many other server options we have.

BPplays committed:
brought server starting into different func and added enable v6 and v4

BPplays committed:
added error checking for disabling both v6 and v4. added option to prefer v6 for dns. added that stuff to the default config

BPplays committed:
fixed dumb bug

BPplays committed:
changed to settings named disable ipvx

BPplays commented:
@Cohee1207 i've added most of the stuff with the exception of having ipv6 opt-in i don't feel it should be, i think it should just be set up in a way where it doesn't cause any issues if you only have v4, but im open to your opinion


~im also having problems with error handling when i have just my v6 or v4 address in use it errors like this even though the whole server creation is in a try catch do you know why?~ i think i fixed it now

BPplays committed:
fixed failed ips still showing as listening

BPplays committed:
fixed error handling

Cohee1207 commented on default/config.yaml:
```diff
@@ -4,6 +4,12 @@ dataRoot: ./data
+# Enables IPv4
+disableIPv4: false
+# Prefers IPv6 for dns
+preferIPv6: false
```
A config value should be more descriptive, like "dnsPreferIPv6"

Cohee1207 commented on default/config.yaml:
```diff
@@ -4,6 +4,12 @@ dataRoot: ./data
 # Listen for incoming connections
 listen: false
+# Enables IPv6
+disableIPv6: false
```
In my opinion these options would better be reversed and grouped into object, i.e.
```yaml
protocol:
    ipv4: true
    ipv6: false
```

BPplays replied:
i've changed it to be like:
```yaml
protocol:
    ipv4: true
    ipv6: true
```
and changed `preferIPv6` to `dnsPreferIPv6`

BPplays committed:
changed ip protocol config layout

BPplays committed:
small const name changes

BPplays committed:
fixed no error if only available protocol fails, and changed wording of some errors

BPplays committed:
fixed error handling saying 'non-fatal error' for protocol fail even when it's the only one enabled

BPplays committed:
moved more logic to listen error handler

BPplays committed:
fixed eslint issues

BPplays committed:
added more info on when to prefer ipv6 for dns

BPplays commented:
@Cohee1207 i've changed the settings to be like:
```yaml
protocol:
    ipv4: true
    ipv6: true
```
and changed `preferIPv6` to `dnsPreferIPv6`

BPplays committed:
in conf changed one 'ipv6' to 'IPv6' for consistency

Cohee1207 commented on default/config.yaml:
```diff
@@ -4,6 +4,12 @@ dataRoot: ./data
+# Enables IPv6 and/or IPv4
+protocol:
+    ipv4: true
+    ipv6: true
```
ipv6 should be disabled by default for backward compatibility

BPplays replied:
if you mean you think it'll cause issues i've tested with wsl2 with ipv6 disabled completely and it didn't seem to cause any major ones (just a non-fatal error about v6), all it should do if v6 isn't available is not start v6 and give a non-fatal error.

if you mean to keep it consistent with past behavior i don't think we should do that here, but it's your decision so i will if you really want to.
a compromise could be to set it up so a fresh install gets v6 by default but existing installs have it disabled, that way it's the new default but it'll behave like it did before for existing installs

Cohee1207 replied:
There should be no behavioral changes past this PR, even with non-fatal errors.

BPplays replied:
@Cohee1207 this reply is a bit confusingly worded but i assume it means no to it being the default at all even just for new installs.
i hope you'll reconsider sometime, but I'll go ahead and make it disabled by default. 
also we could try to detect if the system has v6 and only try to start with v6 if it does

Cohee1207 commented on server.js:
```diff
@@ -54,8 +54,9 @@ if (process.versions && process.versions.node && process.versions.node.match(/20
 
-// Set default DNS resolution order to IPv4 first
-dns.setDefaultResultOrder('ipv4first');
+
```
You can remove these empty lines

Cohee1207 commented on server.js:
```diff
@@ -138,6 +148,33 @@ const enableAccounts = getConfigValue('enableUserAccounts', DEFAULT_ACCOUNTS);
+
+
+if (!enableIPv6 && !enableIPv4) {
+    console.error('error: you can\'t disable all internet you must enable at least ipv6 or ipv4');
```
This wording is weird. Consider the following:

`You can't disable all protocols: at least IPv6 or IPv4 should be enabled.`

BPplays replied:
i've changed it to something similar:

`You can't disable all internet protocols: at least IPv6 or IPv4 must be enabled.`

steve02081504 commented on server.js:
```diff
@@ -708,22 +879,5 @@ userModule.initUserStorage(dataRoot)
-                tavernUrl.hostname,
-                postSetupTasks,
-            );
-        }
+        startServer();
     });
```
```suggestion
    .finally(startServer);
```

Cohee1207 commented on server.js:
```diff
@@ -674,6 +727,124 @@ function logSecurityAlert(message) {
+                key: fs.readFileSync(cliArguments.keyPath),
+            }, app);
+
+        ipv6Server.on('error', (error) => {
```
I don't think that handling async errors like this will work, since you're returning boolean flags immediately, and checking them before the server starts listening or the error emission.

Wrap server creation in a Promise.

```js
function createHttpServer(url) {
    return new Promise((resolve, reject) => {
        const server = http.createServer(app);
        server.on('error', reject);
        server.on('listening', resolve);
        server.listen(url.port, url.hostname);
    }
}

let v4failed = false;

if (enableIPv4) {
    try {
        await createHttpServer(tavernUrl);
    } catch {
        v4failed = true;
    }
}
```

Source: ['listening'](https://nodejs.org/api/net.html#event-listening)

BPplays replied:
@Cohee1207 i've changed to using this method.

Cohee1207 replied:
Thanks. I'll review and test it shortly and will merge it by Sunday if no more fixes are required.

BPplays committed:
changed error message and redid how starting the server works

BPplays committed:
removed unneeded log

BPplays committed:
removed unneeded log

BPplays committed:
removed unneeded comments

BPplays committed:
fixed errors

BPplays committed:
fixed errors

BPplays committed:
fixed errors

BPplays committed:
changed the wording of ip related error messages

BPplays committed:
removed empty lines

BPplays committed:
changed to .finally(startServer);

BPplays committed:
removed some whitespace

BPplays committed:
disabled ipv6 by default ╯︿╰ and changed some message wording

Cohee1207 committed:
Merge branch 'staging' into ipv6

Cohee1207 commented:
Problem: the address displayed in the console for listen mode in IPv6 is not valid in browsers
```
http://[::]:8002/
```

Cohee1207 commented on server.js:
```diff
@@ -554,7 +590,7 @@ const tavernUrl = new URL(
 const autorunUrl = new URL(
     (cliArguments.ssl ? 'https://' : 'http://') +
-    ('127.0.0.1') +
+    ('localhost') +
```
This is going to break localstorage and other domain-specific settings users already have set

BPplays commented:
> Problem: the address displayed in the console for listen mode in IPv6 is not valid in browsers
> 
> ```
> http://[::]:8002/
> ```

it doesn't seem to have given a valid link before:

here is the log for what it's listening on for `staging`
```
    console.log(color.green('SillyTavern is listening on: ' + tavernUrl));
```


here is `tavernUrl` if you enable listen it is 0.0.0.0 which also is not a valid link i just copied this behavior, it also doesn't tell users to go here
```
const tavernUrl = new URL(
    (cliArguments.ssl ? 'https://' : 'http://') +
    (listen ? '0.0.0.0' : '127.0.0.1') +
    (':' + server_port),
);
```








here is a log from staging:
```
--> .\Start.bat --listen

> sillytavern@1.12.4 postinstall
> node post-install.js


up to date in 2s
Node version: v20.16.0. Running in production environment.
Using data root: ./data


SillyTavern 1.12.4
Running 'staging' (40ee236c) - 2024-08-01 01:34:49 +0300

Launching...
SillyTavern is listening on: http://0.0.0.0:8000/

0.0.0.0 means SillyTavern is listening on all network interfaces (Wi-Fi, LAN, localhost). If you want to limit it only to internal localhost (127.0.0.1), change the setting in config.yaml to "listen: false". Check "access.log" file in the SillyTavern directory if you want to inspect incoming connections.
```

Cohee1207 commented:
However, `http://0.0.0.0:8000/` is a valid link and will open a requested page since it's a fully 100% valid URL. Try it out.

BPplays commented:
@Cohee1207 doesn't seem to work on Firefox nightly at least if you paste it in the URL bar. i can probably make a separate message for where to go because i think what it's listening on is still valuable to show

![a](https://raw.githubusercontent.com/BPplays/images/main/pr/st_ipv6/00deb/localhost.webp)

![a](https://raw.githubusercontent.com/BPplays/images/main/pr/st_ipv6/00deb/0-0-0-0.webp)

Cohee1207 commented:
Works for me.

<img width="734" alt="image" src="https://github.com/user-attachments/assets/7f212291-178f-4c93-9745-3af2ccf9e22f">

BPplays committed:
added auto mode for autorun hostname and changed formatting for listening log and added goto message with autorun url

BPplays committed:
added autorun port override

BPplays committed:
removed debug log

BPplays commented:
@Cohee1207 i changed the autorun url to be configurable and default to an automatic mode that choses based on what protocols you have enabled, added an option to override the autorun port, and a message telling the user to go to the autorun url

BPplays committed:
changed formatting

BPplays committed:
added cli args to ipv6 and autorun stuff

BPplays committed:
moved cli args around

Cohee1207 commented on default/config.yaml:
```diff
@@ -4,15 +4,26 @@ dataRoot: ./data
 # Server port
 port: 8000
+# overrides the port for autorun with open your browser with this port and ignore what port the server is running on. -1 is use server port
+autorunPortOverride: -1
```
What's the use case for this? I can't see why this setting could be used.

BPplays replied:
@Cohee1207 since i added an option to change the hostname which could be used to point to any hostname or domain i added the port override if you want autorun to open a reverse proxy, e.g. you have a reverse proxy set up for st.example.com port 443 to go to st.internal.example.org port 8000

BPplays committed:
changed formatting

BPplays committed:
changed colors for ip

BPplays committed:
added avoidLocalhost cli arg

BPplays committed:
changed formatting

Cohee1207 commented on server.js:
```diff
@@ -108,6 +137,10 @@ const cliArguments = yargs(hideBin(process.argv))
         type: 'string',
         default: null,
         describe: 'Root directory for data storage',
+    }).option('avoidLocalhost', {
```
Is this exclusively a CLI argument? Shouldn't it be present in config.yaml as well?

Cohee1207 commented on server.js:
```diff
@@ -606,20 +663,83 @@ const preSetupTasks = async function () {
+        logListen += ipv4Color(' IPv4: ' + tavernUrl);
+    }
+
+    let goToLog = 'Go to: ' + autorunColor(autorunUrl) + ' to open SillyTavern';
```
I don't think that you should log both tavernUrl and autorunUrl. It only creates more confusion.

Cohee1207 replied:
One possible solution: log tavernUrl without a protocol, so it would not look like a link that you could (and should follow).

Cohee1207 commented on server.js:
```diff
@@ -606,20 +663,83 @@ const preSetupTasks = async function () {
-    console.log(color.green('SillyTavern is listening on: ' + tavernUrl));
+
+    let ipv6Color = color.green;
+    let ipv4Color = color.red;
```
Why is ipv4 a red color? It is meant to indicate critical errors, etc.

BPplays replied:
@Cohee1207 i just used what [ipvfoo](https://github.com/pmarks-net/ipvfoo) does, i can change it if you want something different

Cohee1207 replied:
Just leave both green. ipvfoo clearly made if favor of IPv6, but we don't do any favoritism.

Cohee1207 replied:
Fixed it myself.

LenAnderson commented:
> Works for me.
> 

0.0.0.0 is a valid *listening* address ("accept any IP pointing to this host"), but not a valid *destination* address.
https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml

Apparently macOS and some linux distros handle this incorrectly and treat 0.0.0.0 destinations as a loopback address.

BPplays committed:
changed to not print protocol on listening

BPplays committed:
added config option for avoid localhost and changed formatting of messages

BPplays committed:
fixed avoid localhost config option

Cohee1207 committed:
Merge branch 'staging' into ipv6

Cohee1207 committed:
Fix ipv4 color

Cohee1207 reviewed (approved):
No more issues with the code from my end, I think we're good to go here. Small cosmetic lints could be resolved post-merge.
